pkg/controllers/securitypolicy_associated_controller.go

```diff
@@ -0,0 +1,241 @@
+		return
+	}
+
+	shouldReconcile := false
+	for _, pod := range podList.Items {
+		if checkPodHasNamedPort(pod, "update") {
+			shouldReconcile = true
+			break
+		}
+	}
+	if !shouldReconcile {
+		log.Info("no pod in namespace is relevant", "namespace", obj.Namespace, "name", obj.Name)
+		return
+	}
+
+	err = reconcileSecurityPolicy(e.Client, podList.Items, l)
+	if err != nil {
+		log.Error(err, "Failed to reconcile security policy")
+	}
+}
+
+type EnqueueRequestForPod struct {
+	Client client.Client
+}
+
+func (e *EnqueueRequestForPod) Create(createEvent event.CreateEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(createEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Update(updateEvent event.UpdateEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(updateEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Delete(deleteEvent event.DeleteEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(deleteEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Generic(genericEvent event.GenericEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(genericEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Raw(evt interface{}, q workqueue.RateLimitingInterface) {
+	var pods []v1.Pod
+	var obj client.Object
+	var oldObj client.Object
+
+	switch et := evt.(type) {
+	case event.CreateEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.UpdateEvent:
+		obj = et.ObjectNew.(*v1.Pod)
+		oldObj = et.ObjectOld.(*v1.Pod)
+		pods = append(pods, *oldObj.(*v1.Pod))
+	case event.DeleteEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.GenericEvent:
+		obj = et.Object.(*v1.Pod)
+	default:
+		log.Error(nil, "Unknown event type", "event", evt)
+	}
+	pod := obj.(*v1.Pod)
+	pods = append(pods, *pod)
+	err := reconcileSecurityPolicy(e.Client, pods, q)
+	if err != nil {
+		log.Error(err, "Failed to reconcile security policy")
+	}
+}
+
+func getAllPodPortNames(pods []v1.Pod) sets.String {
+	podPortNames := sets.NewString()
+	for _, pod := range pods {
+		for _, container := range pod.Spec.Containers {
+			for _, port := range container.Ports {
+				if port.Name != "" {
+					podPortNames.Insert(port.Name)
+				}
+			}
+		}
+	}
+	return podPortNames
+}
+
+func reconcileSecurityPolicy(client client.Client, pods []v1.Pod, q workqueue.RateLimitingInterface) error {
+	podPortNames := getAllPodPortNames(pods)
+	log.V(1).Info("pod named port", "podPortNames", podPortNames)
+	spList := &v1alpha1.SecurityPolicyList{}
+	err := client.List(context.Background(), spList)
+	if err != nil {
+		log.Error(err, "failed to list security policy")
+		return err
+	}
+
+	for _, securityPolicy := range spList.Items {
+		shouldReconcile := false
+		for _, rule := range securityPolicy.Spec.Rules {
+			for _, port := range rule.Ports {
+				if port.Port.Type == intstr.String {
+					if podPortNames.Has(port.Port.StrVal) {
+						shouldReconcile = true
+					}
+				}
+			}
+		}
+		if shouldReconcile {
+			log.Info("reconcile it", "namespace", securityPolicy.Namespace, "name", securityPolicy.Name)
+			q.Add(reconcile.Request{
```

Comment:
So this `q` is the work queue for security policy CR, right? I have this question because the enqueued info only includes the CR's name and namespace, I'm not sure how the reconciler knows the resource type of the equeued object.

Reply:
This is a good question, I think it is like our main logic `Reconcile` in security_controller.go, one queue maps to a specific CR, `obj := &v1alpha1.SecurityPolicy{}` here stands for security policy.